### ISSUE TITLE
Multiline strings are ugly after dumping

### ISSUE THREAD

**Issue by neumond** [yaml/pyyaml#240]
```
>>> lines = """
... a
... b
... c
... """
>>> lines
'\na\nb\nc\n'
>>> print(yaml.dump({'a': lines}, default_flow_style=False))
a: '

  a

  b

  c

  '

>>> # although roundtrip of dump-load is correct
>>> yaml.load(yaml.dump({'a': lines}))
{'a': '\na\nb\nc\n'}
```
This could output much shorter
```
a: |
  a
  b
  c
```

**Comment by ysaakpr** [comment]
```python
def str_presenter(dumper, data):
    try:
        dlen = len(data.splitlines())
        if (dlen > 1):
            return dumper.represent_scalar('tag:yaml.org,2002:str', data, style='|')
    except TypeError as ex:
        return dumper.represent_scalar('tag:yaml.org,2002:str', data)
    return dumper.represent_scalar('tag:yaml.org,2002:str', data)
```

Tried adding this using `add_representer`, But somehow the results are not always consistent. For some, I am able them to get properly block quoted, but for some string, though they are a multi line, they will continue `a: "a\nb\nc\n"` style. 

If anyone can give me a hint, on where I have to make the changes, I can put some amount of time to make the PR

**Comment by ysaakpr** [comment]
My issue was related to https://github.com/yaml/pyyaml/issues/121 its painful to do spend lots of time to findout what is the reason

**Comment by schollii** [comment]
Nice. If there is fundamental issue we could check in another language that has yaml load/dump (like golang) and supports it. Or ruamel-yaml.

**Comment by melezhik** [comment]
Actually, this issue is only reason why I can't use pyyaml on my project, for me changing from `| multiline` to `line1 \n line2` format is not acceptable as it will break readable representation for people using GitOps manifests in our repository, where helm values passed as YAML multiline stings, I have to use `ruamel.yaml`  now instead, which has other drawbacks for me ... sigh

**Comment by ingydotnet** [comment]
OK I've added this to https://github.com/yaml/pyyaml/projects/9

So we'll look at that for the next release, though I can't say when that will happen.

**Comment by schollii** [comment]
Sone suggestions : 

Separate concerns of  representation from data otherwise you will end up with a mess of code that is hard to maintain. 

Eg the | symbol is presentation, it is metadata about the layout of the data. Similarly references are metadata. And I think most people would consider Comments to be data in the context of yaml, but they can also be treated as metadata because they provide info about surrounding data or context. 

So at load time the metadata data must be loaded and stored, and used at output time.

And as mentioned, there may be third-party open source libs out there eg in Java, Go or Javascript (or even in Python) that have already solved this problem. This is one of the purposes of open source, sharing knowledge. There is no reason to reinvent the wheel here, use them as inspiration.

**Comment by melezhik** [comment]
@schollii i am not sure if you understand the issue. let me repaet. pyyaml parser voluntary changes a yaml markup when doing dump. it shouldn't do this or at least this behavior should be configurable.

our programs are not dependent on presentation layers, people who read and edit yaml files are

**Comment by ingydotnet** [comment]
@melezhik right. Since we are probably adding a better config system in the next release, the rough plan for this is that we add a config option for the format of multiline strings to prefer. Also I suspect it should be easy to configure this with a custom function that can provide that the preference.

**Comment by gunstonej** [comment]
found this on [StackOverflow](https://stackoverflow.com/questions/8640959/how-can-i-control-what-scalar-form-pyyaml-uses-for-my-data) as a quick fix for my requirement: 

```python
parsed = {'fdir_root': '/mnt/c/engDev/git_mf/ipyrun/tests/examples/line_graph_batch',
 'fpth_config': '/mnt/c/engDev/git_mf/ipyrun/tests/examples/line_graph_batch/config-shell_handler.json',
 'title': '# Plot Straight Lines\n### example RunApp',
 'configs': []}

import yaml
def str_presenter(dumper, data):
    """configures yaml for dumping multiline strings
    Ref: https://stackoverflow.com/questions/8640959/how-can-i-control-what-scalar-form-pyyaml-uses-for-my-data"""
    if len(data.splitlines()) > 1:  # check for multiline string
        return dumper.represent_scalar('tag:yaml.org,2002:str', data, style='|')
    return dumper.represent_scalar('tag:yaml.org,2002:str', data)

yaml.add_representer(str, str_presenter)
yaml.representer.SafeRepresenter.add_representer(str, str_presenter) # to use with safe_dum

s = yaml.dump(parsed, indent=2)  # , sort_keys=True)
print(s)

>>> configs: []
>>> fdir_root: /mnt/c/engDev/git_mf/ipyrun/tests/examples/line_graph_batch
>>> fpth_config: /mnt/c/engDev/git_mf/ipyrun/tests/examples/line_graph_batch/config-shell_handler.json
>>> title: |-
>>>   # Plot Straight Lines
>>>   ### example RunApp
```

**Comment by ingydotnet** [comment]
Also https://github.com/yaml/pyyaml/issues/121#issuecomment-1018117110

**Comment by cjw296** [comment]
Slight tweak, better handles strings ending in a newline and might be a bit faster:

```
def str_presenter(dumper, data):
    """configures yaml for dumping multiline strings
    Ref: https://stackoverflow.com/questions/8640959/how-can-i-control-what-scalar-form-pyyaml-uses-for-my-data"""
    if data.count('\n') > 0:  # check for multiline string
        return dumper.represent_scalar('tag:yaml.org,2002:str', data, style='|')
    return dumper.represent_scalar('tag:yaml.org,2002:str', data)
```

**Comment by Sir-Fancy** [comment]
Sorry to necro this, but wanted to save others the headache. This solution _only works_ if you do not have trailing spaces on **any** of your lines. If there is a trailing space somewhere, you'll see the original behavior of the string getting all messed up and "\n"s everywhere. To prevent this from accidentally occurring, you can strip them with this modification to @cjw296's code:

```python
def str_presenter(dumper, data):
    if data.count('\n') > 0:
        data = "\n".join([line.strip() for line in data.splitlines()])  # Remove any trailing spaces, then put it back together again
        return dumper.represent_scalar('tag:yaml.org,2002:str', data, style='|')
    return dumper.represent_scalar('tag:yaml.org,2002:str', data)
```

This behavior took way too long than I'd care to admit to track down.

**Comment by manugarri** [comment]
is there a plan to add this to the library, maybe as a style? its a very common use case and its a shame we have to do these hacks to get proper yaml.

**Comment by nirs** [comment]
> ```python
>         data = "\n".join([line.rstrip() for line in data.splitlines()])  # Remove any trailing spaces, then put it back together again
>         return dumper.represent_scalar('tag:yaml.org,2002:str', data, style='|')
> ```

This converts "|" blocks to "|-". To preserve the style we can use:

```python
        # Remove any trailing spaces messing out the output.
        block = "\n".join([line.rstrip() for line in data.splitlines()])
        if data.endswith("\n"):
            block += "\n"
        return dumper.represent_scalar("tag:yaml.org,2002:str", block, style="|")
```

**Comment by perlpunk** [comment]
Here's a PR for discussion:
* #822

**Comment by nirs** [comment]
This is what we use in ramen to avoid this issue:
https://github.com/RamenDR/ramen/blob/main/test/drenv/yaml.py

**Comment by MajorDallas** [comment]
> This is what we use in ramen to avoid this issue: https://github.com/RamenDR/ramen/blob/main/test/drenv/yaml.py

I found that the line

```python
yaml.representer.SafeRepresenter.add_representer(str, _str_presenter)
```

wasn't working with `safe_dump`, though that could be a quirk specific to `oyaml`. This fixes it:

```python
yaml.add_representer(str, _str_presenter, Dumper=yaml.dumper.SafeDumper)
```

**Comment by LiranYoffe** [comment]
The solutions provided here didn't work for me because some of my lines started with a tab. To fix this, I used the solution suggested above but also replaced all tabs with 4 spaces.
```
def str_presenter(dumper, data):
    if data.count('\n') > 0:
        data = "\n".join([line.rstrip().replace("\t", "    ") for line in data.splitlines()])  # Remove any trailing spaces, then put it back together again
        return dumper.represent_scalar('tag:yaml.org,2002:str', data, style='|')
    return dumper.represent_scalar('tag:yaml.org,2002:str', data)
```
If you need to preserve tabs in the middle of a line, you might consider using regular expressions to replace only the tabs at the beginning of lines.
Hope it helps someone...
